### PR TITLE
add installable feature specification

### PR DESCRIPTION
**Pull request by alex-lotz**
This PR adds an installable feature specification that clusters the ROS Mixed-Port related plugins. Moreover, a more descriptive generation of a comment is added.

### CODE REVIEW

**Comment by NadiaHG** [comment]
There are some missed dependencies at least to ros-model projects. I also get an error installing it:

```
Cannot complete the install because one or more required items could not be found.
  Software being installed: ROS-MixedPort 1.0.0.201908060942 (de.seronet_projekt.mixedport.ros.feature.group 1.0.0.201908060942)
  Missing requirement: ROS-MixedPort 1.0.0.201908060942 (de.seronet_projekt.mixedport.ros.feature.group 1.0.0.201908060942) requires 'org.eclipse.equinox.p2.iu; org.smartmdsd.toolchain.feature.group 0.0.0' but it could not be found
```

there is not another way to define the feature together with the required projects?

**Comment by alex-lotz** [comment]
> There are some missed dependencies at least to ros-model projects. I also get an error installing it:
> 
> ```
> Cannot complete the install because one or more required items could not be found.
>   Software being installed: ROS-MixedPort 1.0.0.201908060942 (de.seronet_projekt.mixedport.ros.feature.group 1.0.0.201908060942)
>   Missing requirement: ROS-MixedPort 1.0.0.201908060942 (de.seronet_projekt.mixedport.ros.feature.group 1.0.0.201908060942) requires 'org.eclipse.equinox.p2.iu; org.smartmdsd.toolchain.feature.group 0.0.0' but it could not be found
> ```
> 
> there is not another way to define the feature together with the required projects?

I have tested the installation of this feature, it works fine on my computer. Of course you need both other tools, the SmartMDSD Toolchain and the ROS Tooling (for installing this feature) as these two are the main dependencies of the feature. I don't know of any other way doing it, and this is the standard way anyways.

**Comment by NadiaHG** [comment]
@alex-lotz could you please specify the eclipse version you are using??

the error I am getting  
`Missing requirement: ROS-MixedPort 1.0.0.201908060942 (de.seronet_projekt.mixedport.ros.feature.group 1.0.0.201908060942) requires 'org.eclipse.equinox.p2.iu; org.smartmdsd.toolchain.feature.group 0.0.0' but it could not be found`

if you check the project dependency is wrong : org.eclipse.equinox.p2.**iu** it should be org.eclipse.equinox.p2.**ui**, something is broken. I tested this with the latest  and the 2018-12 version and I am getting always the same error

File: de.seronet_projekt.mixedport.ros.feature/feature.xml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+   </license>
+
+   <requires>
+      <import feature="org.smartmdsd.toolchain"/>
```
Comment (NadiaHG) [inline de.seronet_projekt.mixedport.ros.feature/feature.xml:56]: I am getting here a warning that says: "cannot not be resolved", however I think this is too generic and the version should be specified.

If I let eclipse to list the required (pressing the button `compute`) imports I get this following list:
```
   <requires>
      <import plugin="org.eclipse.xtext.ide"/>
      <import plugin="org.eclipse.xtext.xbase.ide"/>
      <import plugin="org.antlr.runtime" version="3.2.0" match="perfect"/>
      <import plugin="org.eclipse.xtext"/>
      <import plugin="org.eclipse.xtext.xbase"/>
      <import plugin="org.eclipse.equinox.common" version="3.5.0" match="greaterOrEqual"/>
      <import plugin="org.eclipse.xtext.xbase.lib" version="2.14.0" match="greaterOrEqual"/>
      <import plugin="org.eclipse.xtext.util"/>
      <import plugin="org.eclipse.xtend.lib" version="2.14.0" match="greaterOrEqual"/>
      <import plugin="org.apache.log4j"/>
      <import plugin="org.eclipse.xtext.ui"/>
      <import plugin="org.eclipse.xtext.ui.shared"/>
      <import plugin="org.eclipse.xtext.ui.codetemplates.ui"/>
      <import plugin="org.eclipse.ui.editors" version="3.5.0" match="greaterOrEqual"/>
      <import plugin="org.eclipse.ui.ide" version="3.5.0" match="greaterOrEqual"/>
      <import plugin="org.eclipse.ui"/>
      <import plugin="org.eclipse.compare"/>
      <import plugin="org.eclipse.xtext.builder"/>
      <import plugin="org.eclipse.core.runtime"/>
      <import plugin="org.eclipse.core.resources"/>
      <import plugin="org.eclipse.xtext.generator"/>
      <import plugin="org.eclipse.emf.mwe.utils"/>
      <import plugin="com.google.inject"/>
      <import plugin="org.eclipse.core.commands"/>
      <import plugin="org.eclipse.ui.workbench"/>
      <import plugin="org.xtend.smartsoft.generator"/>
      <import plugin="org.ecore.component.componentDefinition"/>
      <import plugin="org.xtext.component.componentDefinition"/>
      <import plugin="org.ecore.component.seronetExtension"/>
      <import plugin="org.eclipse.emf.edit"/>
      <import plugin="org.ecore.base.mixedport"/>
      <import plugin="org.ecore.base.mixedport.edit"/>
      <import plugin="org.eclipse.emf.ecore"/>
      <import plugin="de.fraunhofer.ipa.componentInterface"/>
      <import plugin="de.fraunhofer.ipa.componentInterface.xtext.ui"/>
      <import plugin="javax.inject"/>
      <import plugin="de.seronet_projekt.ros.interfaces.pool.xtext"/>
      <import plugin="de.seronet_projekt.ros.interfaces.pool.xtext.ide"/>
   </requires>
```

File: de.seronet_projekt.xtend.ROS.generator/src/de/seronet_projekt/xtend/ROS/generator/ROS_Callbacks.xtend [hunk 2]
```diff
@@ -132,15 +132,15 @@ class ROS_Callbacks {
 	«FOR sub: comp.allROSPorts.filter[it.port instanceof RosSubscriber]»
 		void «comp.name»RosPortCallbacks::«sub.name»_cb (const «sub.packageString»::«sub.messageString»::ConstPtr &msg)
 		{
-			//FIXME: implement this upcall method, use "COMP->" to access the component's class members
+			// for implementing this method, you can use the "COMP->" macro to access the component's class members
```
Comment (NadiaHG) [inline de.seronet_projekt.xtend.ROS.generator/src/de/seronet_projekt/xtend/ROS/generator/ROS_Callbacks.xtend:135]: can this be part of a separate commit/PR to be merged together with an update of the documentation https://wiki.servicerobotik-ulm.de/_detail/tutorials:ros:rosgeneratedcallback.png?id=tutorials%3Aros%3Amixed-port-component-ros  ??

**Review by NadiaHG — CHANGES_REQUESTED** [review]

**Comment by alex-lotz** [comment]
@ipa-nhg Hmm, that is strange. First of all, the equinox dependency is not specified anywhere within the ROS Mixedport feature, so I don't understand why it is required in your setup. I suspect it is somehow related to your setup, but it is difficult for me to tell. The SmartMDSD feature is specified as a dependency within the ROS MixedPort feature, so this part of the erorr is correct and is supposed to be like that. You need to install the SmartMDSD Feature at the same time as the ROS MixedPort feature, otherwise it will fail.

File: de.seronet_projekt.xtend.ROS.generator/src/de/seronet_projekt/xtend/ROS/generator/ROS_Callbacks.xtend (hunk 2 above)
Comment (alex-lotz) [reply de.seronet_projekt.xtend.ROS.generator/src/de/seronet_projekt/xtend/ROS/generator/ROS_Callbacks.xtend:135]: really? I don't think that a simple comment update justifies the effort of creating a new PR, this is just wasted time...

**Comment by NadiaHG** [comment]
> @ipa-nhg Hmm, that is strange. First of all, the equinox dependency is not specified anywhere within the ROS Mixedport feature, so I don't understand why it is required in your setup. I suspect it is somehow related to your setup, but it is difficult for me to tell. The SmartMDSD feature is specified as a dependency within the ROS MixedPort feature, so this part of the erorr is correct and is supposed to be like that. You need to install the SmartMDSD Feature at the same time as the ROS MixedPort feature, otherwise it will fail.

yes, I think is something related to the eclipse version or/and my setup, I also checked your and mine code are we are not defining this dependency. I will try on other computer, but the version 2018-12 is right, doesn't it?

**Comment by alex-lotz** [comment]
> > @ipa-nhg Hmm, that is strange. First of all, the equinox dependency is not specified anywhere within the ROS Mixedport feature, so I don't understand why it is required in your setup. I suspect it is somehow related to your setup, but it is difficult for me to tell. The SmartMDSD feature is specified as a dependency within the ROS MixedPort feature, so this part of the erorr is correct and is supposed to be like that. You need to install the SmartMDSD Feature at the same time as the ROS MixedPort feature, otherwise it will fail.
> 
> yes, I think is something related to the eclipse version or/and my setup, I also checked your and mine code are we are not defining this dependency. I will try on other computer, but the version 2018-12 is right, doesn't it?

Correct.

**Comment by alex-lotz** [comment]
@ipa-nhg Btw, what I have found useful for testing is to deactivate the bundle pool within the Eclipse installer for testing new/experimental installations, as otherwise you always run the risk of having old versions of the plugins and the new plugins are not correctly recognized. For real installations, the bundle pool can be activated again...

File: de.seronet_projekt.xtend.ROS.generator/src/de/seronet_projekt/xtend/ROS/generator/ROS_Callbacks.xtend (hunk 2 above)
Comment (mathias-luedtke) [reply de.seronet_projekt.xtend.ROS.generator/src/de/seronet_projekt/xtend/ROS/generator/ROS_Callbacks.xtend:135]: >really? I don't think that a simple comment update justifies the effort of creating a new PR, this is just wasted time...

It is not a waste of time.
In general independent changes should be contained in separate commits.

**Comment by NadiaHG** [comment]
I am following these steps:

1. Clean the eclipse configuration (remove the ~/.p2 and the ~/.eclipse folders)
2. Install eclipse using the installer (Eclipse modeling tools package)
3. Install from updatesites releases SmartSoft toolchain and ROS tooling (with help -> Install new software ...)
4. Import from src this repository (https://github.com/Servicerobotics-Ulm/SeRoNet-Tooling-ROS-Mixed-Port/tree/feature/installable-feature) and the SmartSoft projects "org.ecore.component.seronetExtension", "org.ecore.base.mixedport" and "org.ecore.base.mixedport.edit" that are  missed or out-of-date on the release v3.9
5. Install Xtext in order to let eclipse to build the xtext projects (otherwise the xtend-gen folder are missed)
6. Go to the feature definition to dependencies and press the button "Compute"
7. Menu File -> Export and choose the option Deployable features from Plug-in Development category. Select the Feature and give a path for destination (e.g. ~/Desktop/export)
8. Close eclipse and open it again with an empty workspace
9. Help -> Install new software and add the local folder ~/Desktop/export, choose the feature and press install

I am still getting an error (also on a different machine) related to some generic eclipse packages that we are not including at all. I tried updating the marketplace client, with different eclipse versions,  disabling the bundle pool... I can imagine I should get an error related to the non-released packages (i.e.  "org.ecore.component.seronetExtension", "org.ecore.base.mixedport" and "org.ecore.base.mixedport.edit ) but apparently the problem I have has noting to do with our specific dependencies.

As the add of a (previously non-existing) new project can't break things I could merge this PR (by opening a new issue to review the feature once all the required dependencies are released and available through the update site) . Could you please check the changes I suggested (https://github.com/seronet-project/SeRoNet-Tooling-ROS-Mixed-Port/pull/15#pullrequestreview-271304878) ?

**Comment by alex-lotz** [comment]
@ipa-nhg Step 3 is partially incorrect. As you pointed out in step 4, the release v3.9 is out-of-date. The current reference for the ROS mixed-port is the master branch of SmartMDSD Toolchain Github repo. At the moment I have not yet created a new release (which will be later 3.10), because I still have to complete a few refinements related to the ISW OPC UA Backend extension. So if you want to test the current version, use the master branch. Or you can also wait until I finish my work and do a release later, but it will take me some time.

**Comment by alex-lotz** [comment]
@ipa-nhg and @ipa-mdl please approve this PR as otherwise I cannot continue. I don't see any changes required.

**Comment by NadiaHG** [comment]
Sorry, You said here  https://github.com/seronet-project/SeRoNet-Tooling-ROS-Mixed-Port/pull/15#issuecomment-519505281

> you can also wait until I finish my work and do a release later, but it will take me some time.

I understood we have to wait to merge this til this can be tested and works. Is this the case? 

However there are some pending changes, see https://github.com/seronet-project/SeRoNet-Tooling-ROS-Mixed-Port/pull/15#pullrequestreview-271304878

**Comment by alex-lotz** [comment]
The feature has been duplicate-added by ipa-nhg, so this PR is outdated now and I'll close it.
